### ISSUE TITLE
500 error without stacktrace

### ISSUE THREAD

**Issue by hh** [csc/Hanlon#346]
I've tried centos6 and coreos images and templates (currently on master branch) and I getting a 500 error via the API but I'm not sure where it is coming from. If we catch a 500 error, I'd expect to find a backtrace somewhere but I'm not sure where to look.

```
# hanlon node  
Discovered Nodes
         UUID           Last Checkin  Status               Tags               
3gc9DUOPvEzVXiGB51YcHY  50 sec        B       [1,A,cpus_6,Supermicro,nics_2]  
3mYBYEiYiZjpp1384nWGyW  42 sec        B       [2,A,cpus_6,Supermicro,nics_2]  
4Adqmgk4thTHevXbyH4jbU  24 sec        B       [3,B,cpus_6,Supermicro,nics_2]  
4GITGLQM2bVaL0zmPfcsMC  17 sec        B       [5,C,cpus_6,Supermicro,nics_2]  
4Fs5JcOSvFviyO7tI0EXSS  16 sec        B       [6,C,cpus_6,Supermicro,nics_2]  
5WOalS415Tbf1nZRlHc0Vs  6 sec         B       [4,B,cpus_6,Supermicro,nics_2]  
```

```
# hanlon policy                                                                                                                      
Policies:
#  Enabled   Label   Tags    Model Label    #/Max  Counter           UUID           
0  true     coreos   [A]   install_coreos   2/-    2        54rde0iWYOYKktJliTZCga  
1  true     centos6  [B]   install_coreos   2/-    2        20EWplHwi61nWc87Oeuwlk  
2  true     centos6  [C]   install_centos6  2/-    2        12gZgBJMA09S49LB8P7zMO  
```

```
 # hanlon active_model
Active Models:
 Label   State        Node UUID         Broker  Bind #           UUID           
coreos   init   3gc9DUOPvEzVXiGB51YcHY  none    1       5xwhx15PFJmEtQHa1DOI5A  
coreos   init   3mYBYEiYiZjpp1384nWGyW  none    2       66GjEDjuTfj5R3GGN8uQqO  
centos6  init   4Adqmgk4thTHevXbyH4jbU  none    1       2QKggN7bShj2oN9DiRrdjk  
centos6  init   4Fs5JcOSvFviyO7tI0EXSS  none    2       191764WOo5EMHFLDWTuUmG  
centos6  init   4GITGLQM2bVaL0zmPfcsMC  none    1       17FBci0hCm70JYIzHjVHNg  
centos6  init   5WOalS415Tbf1nZRlHc0Vs  none    2       2jq2wxs1GaWBwx4pNTx4no  
```
```
# curl 'http://1.1.1.11:8026//hanlon/api/v1/node/checkin?uuid=00000000-0000-0000-0000-002590E254EC&mac_id=002590E254EC_002590E254ED&last_state=idle' | jq .
{
  "response": {
    "result": {
      "code": 500,
      "type": "NoMethodError",
      "description": "undefined method `[]' for nil:NilClass"
    }
  }
}
```

**Comment by hh** [comment]
Could be related to #158 or #160 ?

**Comment by tjmcs** [comment]
Issue #158 was all about uncaught errors in the RESTful API, so yes this is related (and #160 was a PR I submitted to fix the problems I knew about). In this case, we're obviously getting a 'nil' when we're expecting an array/hash and then we're trying to parse that array/hash (resulting in the 'internal error' from the RESTful API). This issue arose as a side-effect of the switch we made last year from the RESTful API calling the CLI (which handled and reported errors) to the CLI calling the RESTful API (which does a good job of returning sensible errors most of the time, but sometimes does something like this for some odd edge case).

Given that it's a checkin operation that's returning the error, I'm pretty sure of where to go to find the problem (it's in the Grape code that handles a GET against the /checkin endpoint, but the specific error shown here may actually be drifting up from the engine or an underlying slice object. What would be really helpful would be to understand which node is checking in when you execute that `curl` command (and which active_model instance that node is bound to). It would also be helpful to understand the details of the model in question (CoreOS or CentOS6) so that I can test it locally and see if I'm seeing the same issue here...

**Comment by hh** [comment]
The setup from scratch:
```
#!/bin/bash                                                                                                                                        
hanlon image add -t mk -p ./hnl_mk_debug-image.2.0.1.iso

# Pair 'A'                                                                                                                                         
UUID=$(hanlon tag add -n 1A -t A | grep UUID | awk '{print $3}')
hanlon tag $UUID matcher add --key mk_ipmi_IP_Address --compare equal --value 1.1.0.1

UUID=$(hanlon tag add -n 2A -t A | grep UUID | awk '{print $3}')
hanlon tag $UUID matcher add --key mk_ipmi_IP_Address --compare equal --value 1.1.0.2

# Pair 'B'                                                                                                                                         
UUID=$(hanlon tag add -n 3B -t B | grep UUID | awk '{print $3}')
hanlon tag $UUID matcher add --key mk_ipmi_IP_Address --compare equal --value 1.1.0.3

UUID=$(hanlon tag add -n 4B -t B | grep UUID | awk '{print $3}')
hanlon tag $UUID matcher add --key mk_ipmi_IP_Address --compare equal --value 1.1.0.4

# https://github.com/csc/Hanlon/pull/329                                                                                                           
# http://stable.release.core-os.net/amd64-usr/557.2.0/coreos_production_iso_image.iso                                                              
hanlon image add -t os -p ./coreos_production_iso_image.iso -v  557 -n coreos557
IMAGE_UUID=$(hanlon image | grep coreos | awk '{print $1}')
hanlon model add --template coreos_stable --label install_coreos --image-uuid $IMAGE_UUID -o coreos-model.yml
MODEL_UUID=$(hanlon model | grep coreos | awk '{print $5}')
hanlon policy add --template linux_deploy --label coreos --model-uuid $MODEL_UUID  --broker-uuid=none -e true --tags A

# http://mirrors.cat.pdx.edu/centos/6.6/isos/x86_64/CentOS-6.6-x86_64-minimal.iso                                                                  
hanlon image add -t os -p ./CentOS-6.6-x86_64-minimal.iso -v 6.6 -n centos6
IMAGE_UUID=$(hanlon image | grep centos6 | awk '{print $1}')
hanlon model add --template centos_6 --label install_centos6 --image-uuid $IMAGE_UUID --option centos6.yml
MODEL_UUID=$(hanlon model | grep centos6 | awk '{print $6}')
hanlon policy add --template linux_deploy --label centos6 --model-uuid $MODEL_UUID  --broker-uuid=none -e true --tags B
```

```
[5] pry(#<Grape::Middleware::Error>)> e.backtrace
=> ["/home/hanlon/core/model/base.rb:87:in `fsm_action'",
 "/home/hanlon/core/model/redhat.rb:201:in `mk_call'",
 "/home/hanlon/core/policy/linux_deploy.rb:22:in `mk_call'",
 "/home/hanlon/core/engine.rb:95:in `mk_checkin'",
 "/home/hanlon/web/api/api_node_v1.rb:259:in `block (3 levels) in <class:APIv1>'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:47:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:47:in `block in generate_api_method'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:247:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:247:in `run'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:195:in `block in call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:27:in `block in call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:26:in `catch'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:26:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/head.rb:13:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/builder.rb:153:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:196:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:184:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:152:in `block in call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:96:in `block in recognize'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:68:in `optimized_each'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:95:in `recognize'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:141:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:98:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:33:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:29:in `call'",
 "/home/hanlon/web/app.rb:186:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:66:in `block in call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `each'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/static.rb:124:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/rack_patch.rb:13:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/configuration.rb:82:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:507:in `handle_request'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:375:in `process_client'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:262:in `block in run'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `block in spawn_thread'"]
```

```
[6] pry(#<Grape::Middleware::Error>)> e.backtrace
=> ["/home/hanlon/core/model/base.rb:87:in `fsm_action'",
 "/home/hanlon/core/model/coreos.rb:180:in `mk_call'",
 "/home/hanlon/core/policy/linux_deploy.rb:22:in `mk_call'",
 "/home/hanlon/core/engine.rb:95:in `mk_checkin'",
 "/home/hanlon/web/api/api_node_v1.rb:259:in `block (3 levels) in <class:APIv1>'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:47:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:47:in `block in generate_api_method'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:247:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:247:in `run'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:195:in `block in call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:24:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:27:in `block in call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:26:in `catch'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/error.rb:26:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/middleware/base.rb:18:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/head.rb:13:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/builder.rb:153:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:196:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/endpoint.rb:184:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:152:in `block in call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:96:in `block in recognize'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:68:in `optimized_each'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:95:in `recognize'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:141:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:98:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:33:in `call!'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/grape-0.11.0/lib/grape/api.rb:29:in `call'",
 "/home/hanlon/web/app.rb:186:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:66:in `block in call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `each'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/urlmap.rb:50:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/rack-1.6.0/lib/rack/static.rb:124:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/rack_patch.rb:13:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/configuration.rb:82:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:507:in `handle_request'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:375:in `process_client'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/server.rb:262:in `block in run'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `call'",
 "/usr/local/rvm/gems/ruby-2.2.1@hanlon/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `block in spawn_thread'"]
```

**Comment by hh** [comment]
Getting a similar error (likely different reason) when using https://github.com/chef/hanlon-api-ruby

```
pp Hanlon::Api::Broker::ChefMetal.create({name: 'sweetbroker', plugin:'chef_metal', description: 'Chef Metal Broker \m/', req_metadata_params: {}},{user_description: 'install chef', chef_server_url: 'http://127.0.0.1:8000',client_key: 'pem', node_name: 'web083', install_sh_url: 'http://fobar/install.sh'})
#<Hanlon::Api::Broker::ChefMetal:0x00000001f72500
 @client=
  #<Faraday::Connection:0x00000001f720c8
   @builder=
    #<Faraday::RackBuilder:0x00000001f717b8
     @handlers=[Faraday::Request::UrlEncoded, Faraday::Adapter::NetHttp]>,
   @default_parallel_manager=nil,
   @headers={"User-Agent"=>"Faraday v0.9.1"},
   @options=
    #<struct Faraday::RequestOptions
     params_encoder=nil,
     proxy=nil,
     bind=nil,
     timeout=nil,
     open_timeout=nil,
     boundary=nil,
     oauth=nil>,
   @parallel_manager=nil,
   @params={},
   @proxy=nil,
   @ssl=
    #<struct Faraday::SSLOptions
     verify=nil,
     ca_file=nil,
     ca_path=nil,
     verify_mode=nil,
     cert_store=nil,
     client_cert=nil,
     client_key=nil,
     certificate=nil,
     private_key=nil,
     verify_depth=nil,
     version=nil>,
   @url_prefix=
    #<URI::HTTP:0x00000001f70e30 URL:http://1.1.1.11:8026/hanlon/api/v1>>,
 @resource_path="broker",
 @result=
  {"code"=>500,
   "type"=>"NoMethodError",
   "description"=>"undefined method `name=' for nil:NilClass"}>
```

**Comment by tjmcs** [comment]
can this issue be close, @hh (based on the changes we merged in with PR #351)?

**Comment by hh** [comment]
Done.
